### ISSUE TITLE
Metrics panel on Safari macOS overloads system

### ISSUE THREAD

**Issue by andre4ik3** [cockpit-project/cockpit#17263]
### Explain what happens

1. Open Cockpit, install the metrics plugin if it's not already installed.
2. Let it sit for a bit (in my case a day) to gather metrics data.
3. Open the metrics panel in Safari on macOS and observe as your system melts down and the tab repeatedly crashes.

<details>
<summary>Further meltdown (with an unsupported browser)</summary>
In Orion (a webkit-based browser that is nearly identical to Safari and is currently in Beta), there are currently no RAM/CPU usage protections. So, opening it on Orion causes complete system meltdown until you force quit the browser. I do understand Orion is a completely unsupported web browser, but I just would like to mention it here because that's what I used in the later half of the attached video.
</details>

See this video for the meltdown in action:
https://user-images.githubusercontent.com/62390580/164004988-755f8161-8a45-443d-ab9e-019c8123792c.mov

To summarize: on Safari on Mac, Metrics tab causes system overload with video artifacts and tab crashes.

System Details:
MacBook Pro M1 (Late 2020), macOS Big Sur 11.6.5
Safari 14.1.2

### Version of Cockpit

251.3

### Where is the problem in Cockpit?

Metrics

### Server operating system

Red Hat Enterprise Linux

### Server operating system version

Rocky Linux 8.5 (it's downstream from RHEL)

### What browsers are you using?

Safari macOS, Other

### System log

```shell
This is a client-side issue
```

**Comment by martinpitt** [comment]
Thanks for the video!

Your `top` shows ~ 28% CPU usage and 50% RAM usage.  I can't say whether this is "a lot" and whether it significantly increased when opening the metrics page, but it's certainly far from critical. But even if the JavaScript would be 100% busy for a long time, it wouldn't kill the whole OS.

I am not sure why your whole system (i.e. not just the page, but the browser and even desktop) starts to flicker so heavily -- that smells like some video driver bug in MacOS? Indeed the metrics history takes a while to load and render, but at least on Firefox and Chromium it works reasonably well even with a day's worth of data (and it only shows up to a day by default).

Supposedly the metrics page renders something which triggers that video driver bug, but without us being able to reproduce it (no M1s in our team), working around it is difficult. Does this already happen without any history data, i.e. if cockpit-pcp is not enabled/installed? If not, does it already happen if you enable pcp and start with no data? When exactly does it start to go wrong, e.g. when it renders the first minute of actual data?

**Comment by andre4ik3** [comment]
> Does this already happen without any history data, i.e. if cockpit-pcp is not enabled/installed?

No. It gave me a prompt to install it to enable metrics. Then the empty history page loaded fine.

> If not, does it already happen if you enable pcp and start with no data?

As far as I can tell, no. I enabled it and left it for ~5 minutes to test it out, it rendered those 5 minutes fine, although it was already laggy.

> When exactly does it start to go wrong, e.g. when it renders the first minute of actual data?

I can't really tell, but it looks like it goes wrong as soon as it starts to try to render it. Chromium renders instantly and fine, it's not even that much data. Safari just goes nuts for some reason.

<img width="1193" alt="Metrics in Chromium working fine" src="https://user-images.githubusercontent.com/62390580/164215678-cea6cbe6-fb67-4dbd-bbc6-ea29823b2047.png">

**Comment by garrett** [comment]
I worked around this WebKit bug within the past year by adjusting CSS in Cockpit.

Your version of Cockpit ([251, released on 2021-08-18](https://cockpit-project.org/blog/cockpit-251.html)) is older than the fix. If you upgrade Cockpit to 254 or later (latest release is 267), then it should work properly in Safari.

The bug was in:
- https://github.com/cockpit-project/cockpit/issues/16242

The fix was at:
- https://github.com/cockpit-project/cockpit/commit/2b1e7853e4ebeb9d7b1d2a68cff7c6d21cc047e1

As long as you're using Cockpit 254 (2021-09-29) or later, you should be fine with WebKit-based browsers, including Safari, Orion, GNOME Web, DuckDuckGo (on macOS), and all iPhone and iPad browsers (as they're all Safari at heart).

Meanwhile, if you do want the fix, I'd recommend doing one of the following, depending on your comfort level:
- Use Cockpit's COPR, which has EPEL support: https://copr.fedorainfracloud.org/coprs/g/cockpit/cockpit-preview/
- Use another browser on macOS for now.
- Avoiding the metrics page on WebKit based browsers.

As this is in (RH/Rocky/Alma)EL 8.5, I'll leave this open with the suggestion that we should consider backporting the fix, especially since we do also ship a WebKit-based app ([Cockpit Client on Flathub](https://flathub.org/apps/details/org.cockpit_project.CockpitClient), for Linux desktops) that is also affected by this bug. (It works well with Cockpit >=254. But would suffer the same performance problems with Cockpit < 254, including RHEL and its EL family.)

**Comment by martinpitt** [comment]
Thanks @garrett ! Closing as a duplicate of #16242 then.

Note that CentOS 8 Stream and the upcoming RHEL 8.6 have the fix, and for older releases feel free to take the package from Stream or the COPR.

**Comment by andre4ik3** [comment]
Awesome! I'll upgrade my cockpit via the COPR as you said.
